### ISSUE TITLE
Go CLI distribution

### ISSUE THREAD

**Issue by csantanapr** [openwhisk/openwhisk#577]
Related to new Go-CLI  https://github.com/openwhisk/openwhisk/pull/238

The design for python CLI is to be available in `https://$apihost/cli/download`
A user will install using `$ sudo pip install --upgrade https://new-console.ng.bluemix.net/openwhisk/cli/download`

For Go-CLI we should have a good deprecation period and have both available, making transition smooth.

Have the go cli binaries available at `https://$apihost/cli/go/download`
The `go/download` directory will have the build out per platform and cpu architecture
```
linux\...
mac\...
windows\...
```
Nginx should be configure to list directory content
```
location cli/go/download {
               autoindex on;
        }
```
This allow user to navigate to find the download he wants.

Also something like a website bluemix can have direct links:

Some examples of direct url downloads
`https://$apihost/cli/go/download/mac/amd64/wsk`
`https://$apihost/cli/go/download/windows/amd64/wsk`
`https://$apihost/cli/go/download/linux/386/wsk`
`https://$apihost/cli/go/download/linux/amd64/wsk`

All the builds being generated today:
```
.
├── linux
│   ├── 386
│   │   └── wsk
│   ├── amd64
│   │   └── wsk
│   ├── arm
│   │   └── wsk
│   ├── arm64
│   │   └── wsk
│   ├── mips64
│   │   └── wsk
│   ├── mips64le
│   │   └── wsk
│   ├── ppc64
│   │   └── wsk
│   └── ppc64le
│       └── wsk
├── mac
│   ├── 386
│   │   └── wsk
│   ├── amd64
│   │   └── wsk
│   ├── arm
│   │   └── wsk
│   └── arm64
│       └── wsk
└── windows
    ├── 386
    │   └── wsk.exe
    └── amd64
        └── wsk.exe

```

**Comment by mdeuser** [comment]
@nauerz-ibm previously mentioned that the wsk CLI should have parity with the Bluemix CLI's architecture/CPU support.  The Bluemix CLI download page http://clis.ng.bluemix.net/ui/home.html lists the following platform/architectures:

- Mac OS X 
- Linux 64-bit (need to confirm CPU)
- Windows 64-bit

Bluemix provides an installer for the Mac and Windows and a tgz for Linux.  Currently the OpenWhisk CLI is pretty simple, so perhaps just downloading the binary is sufficient.

**Comment by csantanapr** [comment]
HI @mdeuser Also checkout the Cloud Foundry CLI release page, the offer both Installers and Binaries

I think we can start by offering binaries. but if you take a look closely the binaries are tgz for Linux and Mac, and zip for Windows.
If there is savings in download payload by compressing I would suggest we also do the same and offer a compressed payload to help speed the downloads.

**Comment by mdeuser** [comment]
I'll update the build to create a wsk.exe.zip for each Windows binary, a wks.zip (if that's the Mac custom) for each Mac wsk binary, and a wsk.tar.gz for each Linux wsk binary.  These files will be in addition to the uncompressed files and will be located in the same build output folder as each uncompressed counterpart.

**Comment by rabbah** [comment]
tar.gz will work for mac also.

**Comment by mdeuser** [comment]
ok.  mac binaries will be compress to .tar.gz

**Comment by mdeuser** [comment]
Vincent @houshengbo - You will probably want to use the compressed Go binary files as the downloadable binaries.

**Comment by mdeuser** [comment]
For comments.  Both CF and Bluemix name their CLI binary downloads with additional identification labeling.  Such as `Bluemix_CLI_0.3.3_amd64`, `cf-cli_6.19.0_winx64.zip`, `cf-cli_6.19.0_linux_i686.tgz`

Probably something for us to consider.  We can handle the platform/architecture bits right now (if agreed), but will need some OpenWhisk wide version value (may be there and I just don't know about it) to add the version number.

/cc @csantanapr @houshengbo

**Comment by csantanapr** [comment]
I think it will be good to have granular download os+arch
About the CLI version, today the CLI has a version and it's the build id timestamp

**whisk CLI version	2016-06-09T16:31:10Z**

You get the version from today using `wsk property get` check with @dubeejw where is this value

I think since we should try to keep the url constant across builds, that way people, the UI, the url doesn't change to download it from the nginx proxy `https://$apihost/cli/go/download/mac/amd64/wsk_amd64.tgz`


@rabbah what do you think?

```
$ wsk property get
whisk auth		23bc46b1-...
whisk API host		192.168.99.102
whisk API version	v1
whisk namespace		_
whisk CLI version	2016-06-09T16:31:10Z
```

**Comment by mdeuser** [comment]
Perhaps for another issue, but the CLI version being a timestamp does not seem right to me.  There's a version in the python CLI download file, `openwhisk-0.1.0.tar.gz`, that does not match the `whisk CLI version` property value.  Perhaps what we currently have is more like a "CLI build" - similar to the `whisk API build` property for the API.

**Comment by csantanapr** [comment]
@mdeuser So we are in agreement to not include any build id in to the name of the binary and be just os+arch `wsk_mac_amd64.tgz` `wsk_win_amd64.zip` `wsk_linux_i386.tgz` etc...

**Comment by mdeuser** [comment]
@csantanapr - Yes, I don't think the build id is helpful.  But perhaps a CLI version (v.r.m) would be.

**Comment by csantanapr** [comment]
OK Mark we are in agreement again  the v.r.m doesn't exist today but the in the future will be helpful

**Comment by mdeuser** [comment]
See https://github.com/openwhisk/openwhisk/pull/675#issuecomment-226796231 for current compressed file names.

/cc @houshengbo @johnsolm

**Comment by csantanapr** [comment]
I was able to confirm that https://$APIHOST/cli/go/download/ now contains the go CLIs 👍

**Comment by mdeuser** [comment]
@csantanapr - what was the $APIHOST value you used?

**Comment by rabbah** [comment]
Are these tested yet?

**Comment by mdeuser** [comment]
No.  Test case issue -> #782

**Comment by johnsolm** [comment]
@dubeejw has informed me that https://openwhisk-test.stage1.ng.bluemix.net/cli/go/download is now working properly.  

Waiting on https://openwhisk.stage1.ng.bluemix.net/cli/go/download to be available before release the Whisk UI CLI Download Page.

**Comment by mdeuser** [comment]
Go CLI is now available for download from openwhisk-dev, bluemix YS0, and bluemix YP.  Automated tests are also in place to validate the download from nginx.  I believe this issue can be closed.

@dubeejw @houshengbo @johnsolm - Do you concur that this issue can be closed?

**Comment by dubeejw** [comment]
👍

**Comment by mdeuser** [comment]
@johnsolm @houshengbo - Feel free to open another issue if there is something missing from the Go CLI distribution work.

**Comment by houshengbo** [comment]
I will say this issue can be closed. Thank you, folks, for all the tests.
